dipy/reconst/fwdti.py

```diff
@@ -0,0 +1,412 @@
+    Diso : float, optional
+        Value of the free water isotropic diffusion. Default is set to 3e-3
+        $mm^{2}.s^{-1}$. Please ajust this value if you are assuming different
+        units of diffusion.
+    piterations : inter, optional
+        Number of iterations used to refine the precision of f. Default is set
+        to 3 corresponding to a precision of 0.01.
+    riterations : inter, optional
+        Number of iteration repetitions with adapted S0. To insure that S0 is
+        taken as a model free parameter Each precision iteration is repeated
+        riterations times with adapted S0.
+
+    Returns
+    -------
+    All parameters estimated from the free water tensor model.
+    Parameters are ordered as follows:
+        1) Three diffusion tensor's eigenvalues
+        2) Three lines of the eigenvector matrix each containing the
+           first, second and third coordinates of the eigenvector
+        3) The volume fraction of the free water compartment
+    """
+    W = design_matrix
+
+    # DTI ordinary linear least square solution
+    log_s = np.log(sig)
+    ols_result = np.dot(inv_design, log_s)
+
+    # Define weights as diag(yn**2)
+    S2 = np.diag(np.exp(2 * np.dot(W, ols_result)))
+
+    # DTI weighted linear least square solution
+    WTS2 = np.dot(W.T, S2)
+    inv_WT_S2_W = np.linalg.pinv(np.dot(WTS2, W))
+    WT_S2_LS = np.dot(WTS2, log_s)
+    params = np.dot(inv_WT_S2_W, WT_S2_LS)
+    
+    # General free-water signal contribution
+    fwsig = np.exp(np.dot(design_matrix, 
+                          np.array([Diso, 0, Diso, 0, 0, Diso, -np.log(1.)])))
+
+    df = 1  # initialize precision
+    flow = 0  # lower f evaluated
+    fhig = 1  # higher f evaluated
+    ns = 9  # initial number of samples per iteration
+    nvol = len(sig)
+    for p in range(piterations):
+        df = df * 0.1
+        fs = np.linspace(flow+df, fhig-df, num=ns)  # sampling f
+        # repeat fw contribution for all the samples
+        SFW = np.array([fwsig,]*ns)
+        FS, SI = np.meshgrid(fs, sig)
+        for r in range(riterations):
+            # Free-water adjusted signal
+            S0 = np.exp(-params[6])
+            y = np.log((SI - FS*S0*SFW.T) / (1 - FS))
+
+            # Estimate tissue's tensor from inv(A.T*S2*A)*A.T*S2*y
+            S2 = np.diag(np.square(np.dot(W, params)))
+            WS2 = np.dot(W.T, S2)
+            invWS2W = np.linalg.pinv(np.dot(WS2, W))
+            all_new_params = np.dot(np.dot(invWS2W, WS2), y)
+
+            # compute F2
+            S0r = np.exp(-np.array([all_new_params[6],]*nvol))
+            SIpred = (1-FS)*np.exp(np.dot(W, all_new_params)) + FS*S0r*SFW.T
+            F2 = np.sum(np.square(SI - SIpred), axis=0)
+
+            # Select params for lower F2
+            Mind = np.argmin(F2)
+            params = all_new_params[:, Mind]
+        # Updated f
+        f = fs[Mind]
+        # refining precision
+        flow = f - df
+        fhig = f + df
+        ns = 19
```

Comment:
Agreed, that is an issue with their method and your implementation is very nice. I'm wondering though if it's a practical issue to never evaluate the end points. According to the article, 18% of the measurements are <0.05 and >0.95 so it seems that it would be good to evaluate those to try to avoid getting stuck in a local minimum.